File: chessgpu/tokenizer.py
"""FEN tokenization and the fixed move action space.

Port of DeepMind's `searchless_chess/src/tokenizer.py` and the action-space half
of `src/utils.py` (Apache 2.0). Behaviour is intended to be byte-identical;
`tests/test_tokenizer.py` diffs this against the upstream implementation over
real ChessBench records rather than trusting that claim.

A position becomes exactly 77 tokens over a 31-character vocabulary:

    [  1 ] side to move
    [ 64 ] board, digits expanded to that many '.' fillers
    [  4 ] castling rights, '.'-padded
    [  2 ] en passant square, '..' when none
    [  3 ] halfmove clock, '.'-padded
    [  3 ] fullmove number, '.'-padded

Note the ordering quirk: side-to-move is prepended to the board string, so it
lands at index 0 and the board occupies 1..64. That is upstream's layout and we
match it, because the pretrained checkpoints depend on it.
"""

from __future__ import annotations

import functools

import chess
import numpy as np
from chess import SQUARE_NAMES

# Vocabulary. Order is load-bearing: these indices are what the model learns.
CHARACTERS = [
    "0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
    "a", "b", "c", "d", "e", "f", "g", "h",
    "p", "n", "r", "k", "q",
    "P", "B", "N", "R", "Q", "K",
    "w", ".",
]  # fmt: skip

CHARACTERS_INDEX = {c: i for i, c in enumerate(CHARACTERS)}
VOCAB_SIZE = len(CHARACTERS)          # 31
SEQUENCE_LENGTH = 77

_SPACES = frozenset("12345678")
_PAD = CHARACTERS_INDEX["."]

# Flat 256-entry byte lookup, so tokenize() is dict-free in the hot loop.
_BYTE_TO_INDEX = np.full(256, -1, dtype=np.int16)
for _c, _i in CHARACTERS_INDEX.items():
    _BYTE_TO_INDEX[ord(_c)] = _i


def tokenize(fen: str) -> np.ndarray:
    """Return the 77-token uint8 encoding of a FEN string."""
    board, side, castling, en_passant, halfmoves, fullmoves = fen.split(" ")
    board = side + board.replace("/", "")

    indices: list[int] = []
    for char in board:
        if char in _SPACES:
            indices.extend(int(char) * [_PAD])
        else:
            indices.append(CHARACTERS_INDEX[char])

    if castling == "-":
        indices.extend(4 * [_PAD])
    else:
        indices.extend(CHARACTERS_INDEX[c] for c in castling)
        indices.extend((4 - len(castling)) * [_PAD])

    if en_passant == "-":
        indices.extend(2 * [_PAD])
    else:
        indices.extend(CHARACTERS_INDEX[c] for c in en_passant)

    # Three digits each is enough: the 50-move rule caps halfmoves, and no game
    # reaches move 1000.
    halfmoves += "." * (3 - len(halfmoves))
    indices.extend(CHARACTERS_INDEX[c] for c in halfmoves)

    fullmoves += "." * (3 - len(fullmoves))
    indices.extend(CHARACTERS_INDEX[c] for c in fullmoves)

    if len(indices) != SEQUENCE_LENGTH:
        raise ValueError(
            f"tokenized to {len(indices)} tokens, expected {SEQUENCE_LENGTH}: {fen!r}"
        )
    return np.asarray(indices, dtype=np.uint8)


# --- the same 77 tokens, taken straight off the board ---------------------
#
# `tokenize(board.fen())` is the obvious call and it is the third largest cost
# in the search: 13% of a profiled 5-second run, against 5% for the neural
# network the tokens are for. All of it is a round trip through a string.
# `board.fen()` walks 64 squares building a list of characters, joins them,
# counts runs of empty squares into digits, and glues six fields together with
# spaces; `tokenize` then splits that back apart, expands the digits it just
# created, and looks up every character in a dict.
#
# Nothing in the middle is wanted. The board already holds a bitboard per piece
# type, so the pieces can be scanned out directly -- 32 iterations over the
# occupied squares rather than 64 over every square, with no string built and
# none parsed.
#
# `tokenize` itself is deliberately untouched. It is the function that was
# verified byte-exact against DeepMind's implementation over 12,000 real
# positions, so it stays as the reference, and this is checked against it
# rather than against the paper. `tests/verify_search.py` diffs the two over
# tens of thousands of positions from real games; if they ever disagree, this
# one is wrong.

_SIDE_TOKEN = {chess.WHITE: CHARACTERS_INDEX["w"], chess.BLACK: CHARACTERS_INDEX["b"]}

# FEN writes rank 8 first and python-chess numbers a1 as square 0, so the two
# orders are reversed by rank and agree by file. The +1 is the side-to-move
# token that upstream prepends to the board.
_SQUARE_TO_TOKEN_INDEX = tuple(
    1 + (7 - (square >> 3)) * 8 + (square & 7) for square in range(64)
)

# (token, piece type, colour) for all twelve pieces, resolved once.
_PIECES = tuple(
    (
        CHARACTERS_INDEX[
            chess.piece_symbol(piece_type).upper()
            if colour == chess.WHITE
            else chess.piece_symbol(piece_type)
        ],
        piece_type,
        colour,
    )
    for piece_type in chess.PIECE_TYPES
    for colour in (chess.WHITE, chess.BLACK)
)

# Field offsets, in the layout documented at the top of this module.
_CASTLING_AT, _EP_AT, _HALFMOVE_AT, _FULLMOVE_AT = 65, 69, 71, 74


def tokenize_board(board: chess.Board) -> np.ndarray:
    """The 77-token encoding of `board`, identical to `tokenize(board.fen())`."""
    tokens = [_PAD] * SEQUENCE_LENGTH
    tokens[0] = _SIDE_TOKEN[board.turn]

    for token, piece_type, colour in _PIECES:
        for square in chess.scan_reversed(board.pieces_mask(piece_type, colour)):
            tokens[_SQUARE_TO_TOKEN_INDEX[square]] = token

    # `castling_xfen` is what `Board.fen()` itself calls, so this inherits its
    # handling of Chess960 and of rights that are present but unusable.
    castling = board.castling_xfen()
    if castling != "-":
        for offset, char in enumerate(castling):
            tokens[_CASTLING_AT + offset] = CHARACTERS_INDEX[char]

    # Also matching `Board.fen()`: the square is only part of the position when
    # the capture is actually available. `has_legal_en_passant` generates
    # moves, so it is guarded by the cheap test it would do anyway.
    if board.ep_square is not None and board.has_legal_en_passant():
        name = SQUARE_NAMES[board.ep_square]
        tokens[_EP_AT] = CHARACTERS_INDEX[name[0]]
        tokens[_EP_AT + 1] = CHARACTERS_INDEX[name[1]]

    for at, number in (
        (_HALFMOVE_AT, board.halfmove_clock),
        (_FULLMOVE_AT, board.fullmove_number),
    ):
        digits = str(number)
        if len(digits) > 3:
            # Three characters is all the format has. `tokenize` raises on the
            # same input rather than silently overflowing into the next field,
            # and so does this.
            raise ValueError(f"{number} does not fit the 3-digit field: {board.fen()}")
        for offset, char in enumerate(digits):
            tokens[at + offset] = CHARACTERS_INDEX[char]

    return np.asarray(tokens, dtype=np.uint8)


@functools.lru_cache(maxsize=1)
def _action_space() -> tuple[dict[str, int], tuple[str, ...]]:
    """All moves any piece could make on an empty board, plus promotions.

    A queen and a knight placed on a square between them cover every direction
    any piece moves, so the union over all 64 squares is a superset of every
    legal move in any position. Promotions are enumerated separately since they
    carry a suffix character.
    """
    moves: list[str] = []
    board = chess.BaseBoard.empty()
    for square in range(64):
        targets = []
        board.set_piece_at(square, chess.Piece.from_symbol("Q"))
        targets += board.attacks(square)
        board.set_piece_at(square, chess.Piece.from_symbol("N"))
        targets += board.attacks(square)
        board.remove_piece_at(square)
        for target in targets:
            moves.append(chess.square_name(square) + chess.square_name(target))

    # Promotions. Order matters and mirrors upstream exactly: straight push
    # first, then the capture to the lower file, then to the higher file.
    files = "abcdefgh"
    pieces = ("q", "r", "b", "n")
    promotions: list[str] = []
    for rank, next_rank in (("2", "1"), ("7", "8")):
        for i, file in enumerate(files):
            promotions += [f"{file}{rank}{file}{next_rank}{p}" for p in pieces]
            if file > "a":
                lower = files[i - 1]
                promotions += [f"{file}{rank}{lower}{next_rank}{p}" for p in pieces]
            if file < "h":
                higher = files[i + 1]
                promotions += [f"{file}{rank}{higher}{next_rank}{p}" for p in pieces]
    moves += promotions

    move_to_action: dict[str, int] = {}
    for action, move in enumerate(moves):
        if move in move_to_action:
            raise AssertionError(f"duplicate move in action space: {move}")
        move_to_action[move] = action
    return move_to_action, tuple(moves)


MOVE_TO_ACTION: dict[str, int] = _action_space()[0]
ACTION_TO_MOVE: tuple[str, ...] = _action_space()[1]
NUM_ACTIONS = len(ACTION_TO_MOVE)

# The same mapping, keyed on a move's parts rather than on its UCI string.
#
# `MOVE_TO_ACTION[move.uci()]` is how the search turns a legal move into a
# column of the policy's output, and it does it for every legal move at every
# node it expands: 510,000 calls to `Move.uci()` in a profiled 5-second search,
# 4% of the whole thing, every one of them building a four-character string and
# hashing it only to throw it away.
#
# A move is already three small integers. Packing them into one and indexing a
# flat list skips the string and the hash entirely. 16 bits is exactly enough:
# 6 for the origin square, 6 for the destination, 4 for the promotion piece
# type (0 when there is none).
ACTION_BY_MOVE_KEY: list[int] = [-1] * (1 << 16)


def move_key(move: chess.Move) -> int:
    """Pack a move into the index `ACTION_BY_MOVE_KEY` is keyed on."""
    return move.from_square | (move.to_square << 6) | ((move.promotion or 0) << 12)


for _action, _uci in enumerate(ACTION_TO_MOVE):
    ACTION_BY_MOVE_KEY[move_key(chess.Move.from_uci(_uci))] = _action


def centipawns_to_win_probability(centipawns: float) -> float:
    """Lichess's accuracy curve. See https://lichess.org/page/accuracy."""
    return 0.5 + 0.5 * (2 / (1 + np.exp(-0.00368208 * centipawns)) - 1)

File: rust/src/tokenizer.rs
//! The 77-token encoding, ported from `chessgpu/tokenizer.py::tokenize_board`.
//!
//! # Why, with a correction to the reason
//!
//! This file was written on the premise that a FEN-based FFI contract would force
//! Python to call `tokenize(fen)` at **37.3 us** against `tokenize_board`'s
//! **9.6 us**, making the port's value path slower than the engine it replaces.
//! **That premise was wrong**, and the benchmark said so:
//!
//! ```text
//!   python tokenize(fen)          6.86 us/pos
//!   python tokenize_board(board)  8.28 us/pos
//!   rust,  batched + frombuffer   0.55 us/pos     16x
//! ```
//!
//! The documented 37.3 us is for `tokenize(board.fen())` -- it INCLUDES generating
//! the FEN, which this port already does in Rust. Given a FEN string in hand,
//! parsing it is if anything cheaper than reading the bitboards from Python. So
//! there was never a 4x regression to avoid.
//!
//! The plain reason stands on its own: **16x**, about 8 us per position, and both
//! the policy and value nets tokenize every leaf, so it is paid twice per node.
//! Roughly 2.6% of the Python engine's wall clock.
//!
//! The lesson is the one this port keeps relearning: a number quoted from notes
//! carries its measurement conditions with it, and 37.3 vs 6.86 is the difference
//! between "required" and "worth 2.6%".
//!
//! # Bit-exactness here is easy, and that is worth saying
//!
//! Unlike the prior softmax -- which is unportable because numpy's float32 `exp`
//! is not correctly rounded -- this is integer and character work. There is no
//! rounding, no summation order, no vendor SIMD kernel. Byte-identical output is
//! reachable by construction, and `tests/verify_tokenizer.py` checks it against
//! BOTH Python entry points over real positions.
//!
//! # The two things that are easy to get wrong
//!
//! 1. **The en passant field is gated on legality.** `tokenize_board` writes the
//!    ep square only when `has_legal_en_passant()`, matching `Board.fen()`. This
//!    is the same trap Stage A found; getting it wrong feeds the model a
//!    different position while every move played stays legal.
//! 2. **Rank order is reversed.** FEN writes rank 8 first, python-chess numbers
//!    a1 as square 0. The mapping is `1 + (7 - rank) * 8 + file`, where the `+1`
//!    is the side-to-move token upstream prepends.

use crate::board::{scan_reversed, square_file, square_rank, Board, Color, PieceType};

/// Vocabulary. **Order is load-bearing**: these indices are what the model
/// learnt, so a reordering silently changes every input the network has ever
/// seen. Transcribed from `tokenizer.CHARACTERS` and checked by
/// `tests/verify_tokenizer.py`.
///
/// Note `b` appears only once, in the file-name group at index 11, and serves as
/// file b, black-to-move, and the black bishop. That is upstream's choice and it
/// is not a bug.
pub const CHARACTERS: [u8; 31] = *b"0123456789abcdefghpnrkqPBNRQKw.";

pub const VOCAB_SIZE: usize = 31;
pub const SEQUENCE_LENGTH: usize = 77;

/// Field offsets, matching `tokenizer._CASTLING_AT` and friends.
const CASTLING_AT: usize = 65;
const EP_AT: usize = 69;
const HALFMOVE_AT: usize = 71;
const FULLMOVE_AT: usize = 74;

const PAD: u8 = 30; // index of '.'
const SIDE_WHITE: u8 = 29; // index of 'w'
const SIDE_BLACK: u8 = 11; // index of 'b'

/// Byte value -> vocabulary index, or 255 for anything outside the vocabulary.
const fn byte_table() -> [u8; 256] {
    let mut t = [255u8; 256];
    let mut i = 0;
    while i < CHARACTERS.len() {
        t[CHARACTERS[i] as usize] = i as u8;
        i += 1;
    }
    t
}
static BYTE_TO_INDEX: [u8; 256] = byte_table();

#[inline]
const fn token_index(square: u8) -> usize {
    // FEN writes rank 8 first; python-chess numbers a1 as 0. +1 for the
    // side-to-move token.
    1 + (7 - (square >> 3) as usize) * 8 + (square & 7) as usize
}

#[inline]
fn piece_token(pt: PieceType, color: Color) -> u8 {
    let c: u8 = match (pt, color) {
        (PieceType::Pawn, Color::White) => b'P',
        (PieceType::Knight, Color::White) => b'N',
        (PieceType::Bishop, Color::White) => b'B',
        (PieceType::Rook, Color::White) => b'R',
        (PieceType::Queen, Color::White) => b'Q',
        (PieceType::King, Color::White) => b'K',
        (PieceType::Pawn, Color::Black) => b'p',
        (PieceType::Knight, Color::Black) => b'n',
        (PieceType::Bishop, Color::Black) => b'b',
        (PieceType::Rook, Color::Black) => b'r',
        (PieceType::Queen, Color::Black) => b'q',
        (PieceType::King, Color::Black) => b'k',
    };
    BYTE_TO_INDEX[c as usize]
}

impl Board {
    /// The 77 tokens, identical to `tokenizer.tokenize_board(board)`.
    ///
    /// Returns `Err` rather than panicking when a move counter will not fit its
    /// three-character field, which is what the Python does: silently overflowing
    /// into the next field would corrupt the castling or ep tokens instead of
    /// reporting the problem.
    pub fn tokenize(&self) -> Result<[u8; SEQUENCE_LENGTH], String> {
        let mut tokens = [PAD; SEQUENCE_LENGTH];

        tokens[0] = if self.turn == Color::White {
            SIDE_WHITE
        } else {
            SIDE_BLACK
        };

        // Pieces. Any iteration order is fine -- each square holds one piece --
        // so this walks the six bitboards rather than upstream's twelve
        // (piece_type, colour) pairs.
        for (pt, bb) in [
            (PieceType::Pawn, self.pawns),
            (PieceType::Knight, self.knights),
            (PieceType::Bishop, self.bishops),
            (PieceType::Rook, self.rooks),
            (PieceType::Queen, self.queens),
            (PieceType::King, self.kings),
        ] {
            for sq in scan_reversed(bb & self.occupied_co[0]) {
                tokens[token_index(sq)] = piece_token(pt, Color::White);
            }
            for sq in scan_reversed(bb & self.occupied_co[1]) {
                tokens[token_index(sq)] = piece_token(pt, Color::Black);
            }
        }

        // Castling, in `castling_xfen` order: high square first per colour, so
        // K then Q then k then q. Verified against python-chess's own ordering.
        let rights = self.clean_castling_rights();
        let mut at = CASTLING_AT;
        for (bit, ch) in [
            (7u8, b'K'),
            (0u8, b'Q'),
            (63u8, b'k'),
            (56u8, b'q'),
        ] {
            if rights & (1u64 << bit) != 0 {
                tokens[at] = BYTE_TO_INDEX[ch as usize];
                at += 1;
            }
        }

        // En passant, ONLY when the capture is actually available. See the
        // module docstring; this is the Stage A trap.
        if let Some(sq) = self.ep_square {
            // The legality test runs move generation, so it is second: most
            // positions have no ep square at all and pay nothing.
            if self.has_legal_en_passant() {
                tokens[EP_AT] = BYTE_TO_INDEX[(b'a' + square_file(sq)) as usize];
                tokens[EP_AT + 1] = BYTE_TO_INDEX[(b'1' + square_rank(sq)) as usize];
            }
        }

        for (at, number) in [
            (HALFMOVE_AT, self.halfmove_clock),
            (FULLMOVE_AT, self.fullmove_number),
        ] {
            let digits = number.to_string();
            if digits.len() > 3 {
                return Err(format!(
                    "{} does not fit the 3-digit field: {}",
                    number,
                    self.to_fen()
                ));
            }
            for (offset, ch) in digits.bytes().enumerate() {
                tokens[at + offset] = BYTE_TO_INDEX[ch as usize];
            }
        }

        Ok(tokens)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn vocabulary_is_the_expected_shape() {
        assert_eq!(CHARACTERS.len(), VOCAB_SIZE);
        assert_eq!(BYTE_TO_INDEX[b'.' as usize], PAD);
        assert_eq!(BYTE_TO_INDEX[b'w' as usize], SIDE_WHITE);
        assert_eq!(BYTE_TO_INDEX[b'b' as usize], SIDE_BLACK);
        // 'b' is shared between the file group and the black bishop.
        assert_eq!(piece_token(PieceType::Bishop, Color::Black), SIDE_BLACK);
    }

    #[test]
    fn startpos_tokenizes_to_the_documented_layout() {
        let b = Board::starting_position();
        let t = b.tokenize().expect("startpos fits");
        assert_eq!(t.len(), SEQUENCE_LENGTH);
        assert_eq!(t[0], SIDE_WHITE);
        // Token 1 is a8, which holds a black rook.
        assert_eq!(t[1], piece_token(PieceType::Rook, Color::Black));
        // Token 64 is h1, a white rook.
        assert_eq!(t[64], piece_token(PieceType::Rook, Color::White));
        // All four castling rights, then no ep, then "0.." and "1..".
        assert_eq!(&t[CASTLING_AT..CASTLING_AT + 4], &[
            BYTE_TO_INDEX[b'K' as usize],
            BYTE_TO_INDEX[b'Q' as usize],
            BYTE_TO_INDEX[b'k' as usize],
            BYTE_TO_INDEX[b'q' as usize],
        ]);
        assert_eq!(&t[EP_AT..EP_AT + 2], &[PAD, PAD]);
        assert_eq!(t[HALFMOVE_AT], BYTE_TO_INDEX[b'0' as usize]);
        assert_eq!(t[FULLMOVE_AT], BYTE_TO_INDEX[b'1' as usize]);
    }

    #[test]
    fn move_counter_overflow_is_reported_not_truncated() {
        let mut b = Board::starting_position();
        b.fullmove_number = 1000;
        assert!(b.tokenize().is_err());
    }
}
